### PR TITLE
Fixes nested code actions in LSP bug

### LINKED ISSUE #71335: [LSP] Introduce local for all occurrences code action only changes on occurence
**Note: Turn on the LSP editor to repro** or just use a Razor file

@akhera99 Looks like the new nested code action path stuff has a bug somewhere, or maybe its just the Introduce Local code action?

From debugging, the path for the second nested action is "Introduce Local" / "Introduce Local for args.First()" / "Introduce Local for all occurences of args.First()", so when looking for which code action to resolve, it finds the first nested child and runs that, rather than finding the next one which is the one the user wanted.

RoslynIntroduceLocal

### PR DESCRIPTION
**Pull request by akhera99**
Fixes #71335

The user selected action was not always matching the action being inserted. The code action path was not properly removing the previous nested item from the list.

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs [hunk 1]
```diff
@@ -273,6 +273,10 @@ private static bool IsCodeActionNotSupportedByLSP(IUnifiedSuggestedAction sugges
                             applicableRange: nestedActionSet.ApplicableToSpan.HasValue
                                 ? ProtocolConversions.TextSpanToRange(nestedActionSet.ApplicableToSpan.Value, documentText) : null,
                             nestedSetNumber, codeActionPath, ref currentHighestSetNumber));
+
+                        // Once CodeActions at this nesting have been added, remove the last element from the code action path
+                        // so that the next nested action can be added to the correct path.
+                        codeActionPath.RemoveAt(codeActionPath.Count - 1);
```
Comment (ToddGrun) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:279]: >RemoveAt [](http://example.com/codeflow?start=39&length=8)

nit: RemoveLast?

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs [hunk 2]
```diff
@@ -273,6 +273,10 @@ private static bool IsCodeActionNotSupportedByLSP(IUnifiedSuggestedAction sugges
                             applicableRange: nestedActionSet.ApplicableToSpan.HasValue
                                 ? ProtocolConversions.TextSpanToRange(nestedActionSet.ApplicableToSpan.Value, documentText) : null,
                             nestedSetNumber, codeActionPath, ref currentHighestSetNumber));
+
+                        // Once CodeActions at this nesting have been added, remove the last element from the code action path
+                        // so that the next nested action can be added to the correct path.
+                        codeActionPath.RemoveLast();
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:279]: i don't quite get it.  shouldn't it that at thee nd of this entire method, we pop off the last item we added?  

Also, sholud the code-action-path be immutable to ensure we're not mutating in place something poitned to by lots of items?

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs [hunk 3]
```diff
@@ -114,30 +113,30 @@ private static bool IsCodeActionNotSupportedByLSP(IUnifiedSuggestedAction sugges
...
 
             AddLSPCodeActions(builder, codeAction, request, codeActionKind, diagnosticsForFix, nestedCodeActionCommand,
-                nestedCodeActions, codeActionPath, suggestedAction);
+                nestedCodeActions, codeActionPathArray, suggestedAction);
 
             return builder.ToArray();
         }
 
-        private static ImmutableArray<LSP.CodeAction> CollectNestedActions(
+        private static (ImmutableArray<LSP.CodeAction> actions, ImmutableArray<string> codeActionPath) CollectNestedActions(
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:139]: why do you need to return the mutated code paction path?

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs [hunk 4]
```diff
@@ -114,30 +113,30 @@ private static bool IsCodeActionNotSupportedByLSP(IUnifiedSuggestedAction sugges
             var diagnosticsForFix = GetApplicableDiagnostics(request.Context, suggestedAction);
 
             using var _ = ArrayBuilder<LSP.CodeAction>.GetInstance(out var builder);
-            var codeActionPathList = ImmutableArray.Create(codeAction.Title);
-            var nestedCodeActions = CollectNestedActions(request, codeActionKind, diagnosticsForFix, suggestedAction, codeActionPathList, isTopLevelCodeAction: true);
+            var codeActionPathList = ImmutableArray<string>.Empty;
+            var (nestedCodeActions, codeActionPath) = CollectNestedActions(request, codeActionKind, diagnosticsForFix, suggestedAction, codeActionPathList, isTopLevelCodeAction: true);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:117]: this is weird.  why are we returning teh path?

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs [hunk 5]
```diff
@@ -113,25 +113,26 @@ private static bool IsCodeActionNotSupportedByLSP(IUnifiedSuggestedAction sugges
             var diagnosticsForFix = GetApplicableDiagnostics(request.Context, suggestedAction);
 
             using var _ = ArrayBuilder<LSP.CodeAction>.GetInstance(out var builder);
-            using var _1 = ArrayBuilder<string>.GetInstance(out var codeActionPathList);
+            var codeActionPathList = ImmutableArray<string>.Empty;
+            codeActionPathList = codeActionPathList.Add(codeAction.Title);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:117]:
```suggestion
            var codeActionPathList = ImmutableArray.Create(codeAction.Title);
```

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs [hunk 6]
```diff
@@ -141,13 +142,17 @@ private static bool IsCodeActionNotSupportedByLSP(IUnifiedSuggestedAction sugges
...
             IUnifiedSuggestedAction suggestedAction,
-            ArrayBuilder<string> codeActionPathList,
+            ImmutableArray<string> codeActionPathList,
             bool isTopLevelCodeAction = false)
         {
             var codeAction = suggestedAction.OriginalCodeAction;
             using var _1 = ArrayBuilder<LSP.CodeAction>.GetInstance(out var nestedCodeActions);
 
-            codeActionPathList.Add(codeAction.Title);
+            if (!isTopLevelCodeAction)
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:151]: why not just always add the title?  i don't really get why we're special casing anything.

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs [hunk 7]
```diff
@@ -224,13 +229,15 @@ private static bool IsCodeActionNotSupportedByLSP(IUnifiedSuggestedAction sugges
...
             ref int currentHighestSetNumber)
         {
             var codeAction = suggestedAction.OriginalCodeAction;
 
             var diagnosticsForFix = GetApplicableDiagnostics(request.Context, suggestedAction);
-            var nestedActions = GenerateNestedVSCodeActions(request, documentText, suggestedAction, codeActionKind, ref currentHighestSetNumber, codeActionPathList);
+            codeActionPathList = codeActionPathList.Add(codeAction.Title);
+            var nestedActions = GenerateNestedVSCodeActions(request, documentText, suggestedAction,
+                codeActionKind, ref currentHighestSetNumber, codeActionPathList);
             var codeActionPath = codeActionPathList.ToArray();
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:241]: inline this.

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs [hunk 8]
```diff
@@ -224,13 +229,15 @@ private static bool IsCodeActionNotSupportedByLSP(IUnifiedSuggestedAction sugges
...
+            ImmutableArray<string> codeActionPathList,
             ref int currentHighestSetNumber)
         {
             var codeAction = suggestedAction.OriginalCodeAction;
 
             var diagnosticsForFix = GetApplicableDiagnostics(request.Context, suggestedAction);
-            var nestedActions = GenerateNestedVSCodeActions(request, documentText, suggestedAction, codeActionKind, ref currentHighestSetNumber, codeActionPathList);
+            codeActionPathList = codeActionPathList.Add(codeAction.Title);
+            var nestedActions = GenerateNestedVSCodeActions(request, documentText, suggestedAction,
+                codeActionKind, ref currentHighestSetNumber, codeActionPathList);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:240]:
```suggestion
            var nestedActions = GenerateNestedVSCodeActions(request, documentText, suggestedAction,
                codeActionKind, ref currentHighestSetNumber, codeActionPathList.Add(codeAction.Title));
```

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs [hunk 9]
```diff
@@ -60,7 +60,7 @@ internal static class CodeActionHelpers
                     {
                         if (!IsCodeActionNotSupportedByLSP(suggestedAction))
                         {
-                            using var _1 = ArrayBuilder<string>.GetInstance(out var codeActionPathList);
+                            var codeActionPathList = ImmutableArray<string>.Empty;
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:63]: delete this.  just pass in `[]` to `codeActionPathList: []` below.

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs [hunk 10]
```diff
@@ -113,25 +112,25 @@ private static bool IsCodeActionNotSupportedByLSP(IUnifiedSuggestedAction sugges
...
             using var _ = ArrayBuilder<LSP.CodeAction>.GetInstance(out var builder);
-            using var _1 = ArrayBuilder<string>.GetInstance(out var codeActionPathList);
+            var codeActionPathList = ImmutableArray.Create(codeAction.Title);
             var nestedCodeActions = CollectNestedActions(request, codeActionKind, diagnosticsForFix, suggestedAction, codeActionPathList, isTopLevelCodeAction: true);
 
             Command? nestedCodeActionCommand = null;
             var title = codeAction.Title;
 
-            var codeActionPath = codeActionPathList.ToArray();
+            var codeActionPathArray = codeActionPathList.ToArray();
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:121]: Don't think you need both `codeActionPathList` and `codeActionPathArray` right?  can probably just inline the ToArray on the places that don't accept an immutable array

**Review by dibarbet — APPROVED** [review]

**Pull request merged**